### PR TITLE
Commit pump

### PR DESCRIPTION
**Pull request by Palakila**
Added Commit Ingestion based on general logic of PullRequest, with a a watermark based on RepoId.

### CODE REVIEW

**Comment by chinarosesz** [comment]
> Copyright (C) 2020 https://github.com/chinarosesz/AzureDevOps.DataIngestor

@mgardnerMS, we should also add the build ingestor here as well when you get a chance.

---
Refers to: README.md:21 in 400d785. [](commit_id = 400d78587ca6f7d3de3ddffa1f65805b847ec889, deletion_comment = False)

File: src/AzureDevOps.DataIngestor.Sdk/Entities/VssCommitEntity.cs [hunk 1]
```diff
@@ -0,0 +1,34 @@
+﻿using System;
+using System.ComponentModel.DataAnnotations;
+using System.ComponentModel.DataAnnotations.Schema;
+
+namespace AzureDevOps.DataIngestor.Sdk.Entities
+{
+    [Table("VssCommit")]
+    public partial class VssCommitEntity : VssBaseEntity
```
Comment (chinarosesz) [inline src/AzureDevOps.DataIngestor.Sdk/Entities/VssCommitEntity.cs:8]: >partial [](http://example.com/codeflow?start=11&length=7)

We probably don't need this to be partial correct?

File: src/AzureDevOps.DataIngestor.Sdk/Ingestors/CommitIngestor.cs [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+            // Query for most recent date from watermark table first
+            DateTime mostRecentDate = this.GetCommitWatermark(repo.Id);
+
+            // Retrieve commits from Azure DevOps
+            List<GitCommitRef> commitLists = await this.vssClient.GitClient.GetCommitsWithRetryAsync(repo.Id, "", mostRecentDate, DateTime.UtcNow);
+
+            if (commitLists.Count > 0)
+            {
+                this.IngestCommits(commitLists, repo);
+                // Update watermark once data is successfully ingested so next time it doesn't repeat
```
Comment (chinarosesz) [inline src/AzureDevOps.DataIngestor.Sdk/Ingestors/CommitIngestor.cs:59]: Line break between these two

File: src/AzureDevOps.DataIngestor.Sdk/Ingestors/CommitIngestor.cs [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+            }
+        }
+
+        private async Task RetrieveAndIngestCommit(TeamProjectReference project, GitRepository repo)
+        {
+            // Query for most recent date from watermark table first
+            DateTime mostRecentDate = this.GetCommitWatermark(repo.Id);
+
+            // Retrieve commits from Azure DevOps
+            List<GitCommitRef> commitLists = await this.vssClient.GitClient.GetCommitsWithRetryAsync(repo.Id, "", mostRecentDate, DateTime.UtcNow);
```
Comment (chinarosesz) [inline src/AzureDevOps.DataIngestor.Sdk/Ingestors/CommitIngestor.cs:54]: >"" [](http://example.com/codeflow?start=110&length=2)

string.empty instead for better code readability.

File: src/AzureDevOps.DataIngestor.Sdk/Ingestors/CommitIngestor.cs [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+            List<TeamProjectReference> projects = await this.vssClient.ProjectClient.GetProjectsAsync(this.projectNames);
+
+            // For each project, retrieve and ingest commit data
+            foreach (TeamProjectReference project in projects)
+            {
+                List<GitRepository> reposFromProject = await this.vssClient.GitClient.GetRepositoriesWithRetryAsync(project.Name);
+
+                foreach (GitRepository repo in reposFromProject)
+                {
+                    this.logger.LogInformation($"Retrieve and ingest commits");
```
Comment (chinarosesz) [inline src/AzureDevOps.DataIngestor.Sdk/Ingestors/CommitIngestor.cs:42]: >Retrieve [](http://example.com/codeflow?start=49&length=8)

Maybe log the project name if it's not logged somewhere else already.

File: src/AzureDevOps.DataIngestor.Sdk/Entities/VssCommitEntity.cs (hunk 1 above)
Comment (Palakila) [reply src/AzureDevOps.DataIngestor.Sdk/Entities/VssCommitEntity.cs:8]: Probably not, I kinda copy from the PR one.

File: src/AzureDevOps.DataIngestor.Sdk/Ingestors/CommitIngestor.cs (hunk 4 above)
Comment (Palakila) [reply src/AzureDevOps.DataIngestor.Sdk/Ingestors/CommitIngestor.cs:42]: It's logged in calls before but does not hurt

File: src/AzureDevOps.DataIngestor.Sdk/Clients/VssGitClient.cs [hunk 5]
```diff
@@ -20,23 +20,38 @@ internal VssGitClient(Uri baseUrl, VssCredentials credentials, ILogger logger) :
...
-            {
-                return await this.GetCommitsAsync(repositoryId, searchCriteria, skip, top);
-            });
+            List<GitCommitRef> commitRefs = new List<GitCommitRef>();
+            try
+            {
+                commitRefs = await RetryHelper.SleepAndRetry(this.retryAfter, this.logger, async () =>
+                {
+                    return await this.GetCommitsAsync(repositoryId, searchCriteria, skip, top);
+                });
```
Comment (chinarosesz) [inline src/AzureDevOps.DataIngestor.Sdk/Clients/VssGitClient.cs:46]: If there are more than 100 commits within the date range, do we need to use a continuation token or retrieve more?

Comment (Palakila) [reply src/AzureDevOps.DataIngestor.Sdk/Clients/VssGitClient.cs:46]: For sure we need more, I missed the 100 default value - as off my screen somehow. I checked the API and don't see a continuationToken. Checking

Comment (Palakila) [reply src/AzureDevOps.DataIngestor.Sdk/Clients/VssGitClient.cs]: Done.

**Review by chinarosesz — APPROVED** [review]
:shipit:

**Pull request merged**